doc/man3/EZ_Clear.3
'\"
'\" Copyright (c) 1997 Maorong Zou
'\" 
.TH EZ_Clear 3 "" EZWGL "EZWGL Functions"
.BS
.SH NAME
EZ_Clear \- clear buffers

.SH SYNOPSIS
.nf
.B #include <EZ.h>
.sp
.BI "void  EZ_Clear(int " which )

.SH ARGUMENTS

\fIwhich\fR Specify which buffer to clear. Must be ORed from
EZ_COLOR_BUFFER_BIT and  EZ_DEPTH_BUFFER_BIT.

.SH DESCRIPTION
.PP
\fBEZ_Clear\fR uses the current clear color to clear the current
frame buffer. If specified, it clears the depth buffer to
its maximum value.


.SH "SEE ALSO"
EZ_ClearColor(3), EZ_ClearIndex(3)
.br




ChangeLog
Fri Dec 10 23:44:16 1999  m z  <zou@hitachi1.ma.utexas.edu>
        
        * Added EZ_SSheetSortRegionR, EZ_SSheetSortColumns/Rows
        
1999-12-04  Maorong Zou  <mzou@linux115.ma.utexas.edu>

        * Released version 1.50 
        * added a CVS repository entry 
        *  :pserver:anonymous@cvs.math.utexas.edu:/home/cvs/EZWGL
        * and a read-only cvs web interface
        *  http://cvs.math.utexas.edu/cgi-bin/cvswe.cgi?cvsroot=EZWGL

1999-11-18  Maorong Zou  <mzou@linux34.ma.utexas.edu>

	* Added THUMB_WHEEN and SPREAD_SHEET widget
	* a few modifications in in FigItem ...

1999-09-22  Maorong Zou  <mzou@linux34.ma.utexas.edu>

	* EZ_RESOURCE_STRING resource
	* EZ_LedLoadPixelMatrix() EZ_LedLoadOnOffMatrix()
	* EZ_AllowDualLabel()
	* EZ_FlashWidgetBorder2()
	* NewFile  EZ_ResourceEdit.c
	* Incompatible change:  EZ_LED_PIXEL_SIZE now
	*        takes a pair of ints (xsize, ysize)

	
Sun Aug 22 01:26:59 1999  m z  <mzou@crab.ma.utexas.edu>

        * Separator Widget

Tue Aug 17 02:10:03 1999  m z  <mzou@crab.ma.utexas.edu>

        * Added   EZ_SlideOpenWidget,  EZ_SlideCloseWidget
        *         EZ_SlideOpenWidgetRelativeTo
        * new widget EZ_WIDGET_HISTOGRAM

Tue Aug 10 12:33:26 1999  m z  <mzou@crab.ma.utexas.edu>

        * Added the pare EZ_FreezeDisplay() and EZ_ThawDisplay()
        * New File EZ_ResourceEdit.c (resource editor functions)
        * Added EZ_RegisterApplicationResources(n,spec,config)
        * Modified all resource handling functions
        * changed EZ_Initialize and EZ_InitializeXrm from void to int
        *   (return -1 on failure, 0 on success)
        
Mon Jul 12 02:13:37 1999  m z  <mzou@crab.ma.utexas.edu>

        * New widget GradientBar
        * Fix a bug in EZ_PixelValue2RGB
        
Tue Jul  6 02:17:29 1999  m z  <mzou@crab.ma.utexas.edu>

        * Fix buffer overflow bug in EZ_Xpm.c
        * added line/word wrap for textWidget, IN BROWSING FILE MODE ONLY!!
        * (need to rewrite textWidget, the current line based
        *  implementation is a design flaw)

Thu Jun 17 01:49:17 1999  m z  <mzou@crab.ma.utexas.edu>

        * New resources: SHAPE_PIXMAP, SHAPE_FILE
        * EZ_Bitmap *EZ_CreateShapeMaskFromFigure(fitem) EZ_Item *fitem;
        * Pixmap EZ_MakePixmapFromFigItem(item, width, height)
             EZ_Item *item; int *width, *height;
        
Fri May 14 23:30:19 1999  m z  <mzou@crab.ma.utexas.edu>

        * Added a new tree type: branched-list-tree

Thu May 13 01:50:55 1999  m z  <mzou@crab.ma.utexas.edu>

        * New Widget:  ScrollButton, SpinButton, Locator

Sun May  9 02:03:17 1999  m z  <mzou@crab.ma.utexas.edu>

        * Added UDF widget resource handling.
        * EZ_RegisterResourcesAll(type, nresrc, resrc, configure);
        * EZ_RegisterResourcesWidget(widget, nresrc, resrc, configure);        

Tue May  4 23:36:31 1999  m z  <mzou@crab.ma.utexas.edu>

        *  Widget (horizontal/vertical) ruler
        *  
        *  Configuration options:
        *  EZ_HIGHLIGHT_BACKGROUND, EZ_HIGHLIGHT_FOREGROUND,
        *  EZ_OVERRIDE_REDIRECT

Tue May  4 01:22:57 1999  m z  <mzou@crab.ma.utexas.edu>

        * New Function 
        * void  EZ_ConstrainFreeLabel(EZ_Widget *widget, int flag);
        * void  EZ_MoveWidgetWindow(EZ_Widget *widget, int x, int y);
        * void  EZ_ResizeWidgetWindow(EZ_Widget *widget, int w, int h);
        * void  EZ_MoveResizeWidgetWindow(EZ_Widget *widget, int x, int y, int w, int h);        
        * Added optimization for manipulating images.
        *
        
Sat Apr 17 01:25:25 1999  m z  <mzou@crab.ma.utexas.edu>

        * Autoconf (finally)

1999-04-16  Maorong Zou  <mzou@linux34.ma.utexas.edu>

        * added a vt100 emulator widget (EZ_WIDGET_TERM), derived from xiterm.
        * added support for image format: png and tiff
        
Tue Dec  8 11:39:04 1998  Maorong Zou  <mzou@linux34.ma.utexas.edu>

        *  new function:     EZ_ScaleRGBT
        *                    EZ_XPixmap2RGB
        *                    EZ_SelectInput
        *                    EZ_CreateBackgroundPixmapsFromRawRGB

Sat Dec  5 03:06:58 1998  m z  <mzou@bob.ma.utexas.edu>

        * EZ_LABEL_SHADOW  "labelShadow" config. option

Thu Dec  3 14:54:21 1998  Maorong Zou  <mzou@linux34.ma.utexas.edu>

        * new function:   EZ_SetDirTreeSorter
        *                 EZ_MarkWidgetDirty
        
Mon Nov 30 12:25:50 1998  Maorong Zou  <mzou@linux34.ma.utexas.edu>

        * EZ_WM_IGNORE  "wmIgnore"    configuration option
        *
        * new function    EZ_GetIconTextGeometry
        *                 EZ_SetGlobalEventFilter(filter, data)
        
Thu Nov 19 16:35:47 1998  Maorong Zou  <mzou@linux34.ma.utexas.edu>

	*  EZ_TILE_ORIGIN  tileOrigin: mode x y
	*  resources.  mode=0,1,2    (0 for off)

	*  It is now possible to make all EZ widgets
	*  pseudo transparent under Enlightment WM ...

	*  A few image readers have been modified to
	*  return RGB images.

	*  new functions EZ_ScaleRGB, EZ_ReadImageFile,
	*                EZ_ScaleLabelPixmap,
	*                EZ_SwapLabelPixmaps, 
	*                EZ_CopyLabelPixmaps,
	*                EZ_TraverseAllWidgets,

	*  In EZ_DisplayWidget, the order for displaying widget tree
	*  is reversed. Now widgets refresh a lot faster visually.
	
	*  Bug Fix:  whe a radio button is destroyed, the widget is
	*            not removed from its button group immediately.



lib/EZ_Clear.c
/*********************************************************************
 *
 *         EZWGL, the EZ Widget and Graphics Library
 *
 * *             Copyright (C) 1996, 1997, 1998  Maorong Zou
 *  
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the Free
 * Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 **********************************************************************/
/*
 *  June 1996.  Beta Release.
 *  Sept 1996.  Release Version 1.0
 *  Dec 1996.  Release Version 1.1 Beta
 *  April 1997.  Release Version 1.2
 *  November 1997.  Release Version 1.3
 *  November 1998.  Release Version 1.4
 *  December 1999.  Release Version 1.50
 */
/********************************************************************
 *
 *  EZ_Clear.c
 */
#include "EZ_Widget.h"
#include "EZ_XImageTypes.h"

/***************************************************************
 *
 *  Functions implemented in this file
 */
void              EZ_Clear_3_XImage00 MY_ANSIARGS((int flag));
void              EZ_Clear_3_XImage MY_ANSIARGS((int flag));
void              EZ_Clear_316_XImage MY_ANSIARGS((int flag));
void              EZ_Clear_332_XImage MY_ANSIARGS((int flag));
void              EZ_Clear_324_XImage MY_ANSIARGS((int flag));


void EZ_Clear_3_XImage00(flag)
     int flag;
{
  unsigned long   pixel_value;
  int              length;
  
  if(flag & EZ_COLOR_BUFFER_BIT)
    {
      if(EZ_ColorMode == EZ_CMAP_MODE)
	pixel_value = EZ_ColorArrayFix[(int)EZ_BackRGBA[0]];
      else 
	pixel_value = RGB2PixelLookUp_Fix(EZ_BackRGBA[0],EZ_BackRGBA[1], EZ_BackRGBA[2]);

      if(pixel_value == 0)
	{
	  length = EZ_Image3->bytes_per_line * EZ_Image3->height;
	  memset(EZ_ImageData, (int)pixel_value, length);
	}
      else
	{
	  int x, y, rows, cols;
	  XImage *ximage = EZ_Image3;
	  rows = EZ_WindowHeight;
	  cols = EZ_WindowWidth;

	  for(y = 0; y < rows; y++) for(x = 0; x < cols; x++)
	    XPutPixel(ximage, x, y, pixel_value);
	}
    }
  if((flag & EZ_DEPTH_BUFFER_BIT) && (EZ_ZBufferSet != 0))
   memset(EZ_Zbuf,0, ((EZ_WindowWidth * EZ_WindowHeight) * sizeof(int)));
}


void EZ_Clear_3_XImage(flag)
     int flag;
{
  unsigned long   pixel_value;
  int             length;
 
  if(flag & EZ_COLOR_BUFFER_BIT)
    {
      if(EZ_ColorMode == EZ_CMAP_MODE)
	pixel_value = EZ_ColorArrayFix[(int)EZ_BackRGBA[0]];  /* EZ_BackRGBA[] are integers */
      else 
	{EZ_DitherRGB(pixel_value,4,5,
		      (int)EZ_BackRGBA[0],
		      (int)EZ_BackRGBA[1], 
		      (int)EZ_BackRGBA[2]);}

      length = EZ_Image3->bytes_per_line * EZ_Image3->height;
      memset(EZ_ImageData, (int)pixel_value, length);
    }
   if((flag & EZ_DEPTH_BUFFER_BIT) && (EZ_ZBufferSet != 0))
     memset(EZ_Zbuf,0, ((EZ_WindowWidth * EZ_WindowHeight)*sizeof(int)));
}

void EZ_Clear_316_XImage(flag)
     int flag;
{
  unsigned long   pixel_value;
  int             m, length;
  unsigned short  *pixptr;
 
  if(flag & EZ_COLOR_BUFFER_BIT)
    {
      if(EZ_ColorMode == EZ_CMAP_MODE)
	pixel_value = EZ_ColorArrayFix[(int)EZ_BackRGBA[0]];
      else 
	pixel_value = RGB2PixelLookUp_Fix(EZ_BackRGBA[0],EZ_BackRGBA[1], EZ_BackRGBA[2]);

      length = EZ_Image3->bytes_per_line * EZ_Image3->height;
      if(pixel_value == 0)
	{
	  memset(EZ_ImageData, (int)pixel_value, length);
	}
      else
	{
	  pixptr = (unsigned short *) EZ_ImageData;
	  for(m = 0; m < (length >> 1); m++)
	    *pixptr++ =(unsigned short )pixel_value;
	}
    }
  if((flag & EZ_DEPTH_BUFFER_BIT) && (EZ_ZBufferSet != 0))
   memset(EZ_Zbuf,0, ((EZ_WindowWidth * EZ_WindowHeight) * sizeof(int)));
}


void EZ_Clear_332_XImage(flag)
     int flag;
{
  unsigned long   pixel_value;
  int             m, length;
  unsigned int   *pixptr;
 
  if(flag & EZ_COLOR_BUFFER_BIT)
    {
      if(EZ_ColorMode == EZ_CMAP_MODE)
	pixel_value = EZ_ColorArrayFix[(int)EZ_BackRGBA[0]];
      else 
	pixel_value = RGB2PixelLookUp_Fix(EZ_BackRGBA[0],EZ_BackRGBA[1], EZ_BackRGBA[2]);
      
      length = EZ_Image3->bytes_per_line * EZ_Image3->height;

      if(pixel_value == 0)
	{
	  memset(EZ_ImageData, (int)pixel_value, length);
	}
      else
	{
	  pixptr = (unsigned int *) EZ_ImageData;
	  for(m = 0; m < (length >> 2); m++)
	    *pixptr++ =(unsigned int )pixel_value;
	}
    }
  if((flag & EZ_DEPTH_BUFFER_BIT) && (EZ_ZBufferSet != 0))
   memset(EZ_Zbuf,0, ((EZ_WindowWidth * EZ_WindowHeight) * sizeof(int)));
}

void EZ_Clear_324_XImage(flag)
     int flag;
{
  unsigned long   pixel_value;
  int             height;
 
  if(flag & EZ_COLOR_BUFFER_BIT)
    {
      if(EZ_ColorMode == EZ_CMAP_MODE)
	pixel_value = EZ_ColorArrayFix[(int)EZ_BackRGBA[0]];
      else 
	pixel_value = RGB2PixelLookUp_Fix(EZ_BackRGBA[0],EZ_BackRGBA[1], EZ_BackRGBA[2]);
      
      height = EZ_Image3->height;

      if(pixel_value == 0)
	{
	  int length = EZ_Image3->bytes_per_line * height;
	  memset(EZ_ImageData, (int)pixel_value, length);
	}
      else
	{
	  int y, x, width;
	  int bpl = EZ_Image3->bytes_per_line;
	  unsigned char *pixptr;
	  unsigned char pv[4];
	  width = EZ_WindowWidth;
	  pixptr = (unsigned char *) &pixel_value;
	  pv[0] = *pixptr++; pv[1] = *pixptr++;
	  pv[2] = *pixptr++; pv[3] = 0;

	  pixptr = (unsigned char *) EZ_ImageData;
	  for(y = 0; y < height; y++, pixptr += bpl)
	    {
	      unsigned char *ptr = pixptr;
	      for(x = 0; x < width; x++)
		{ *ptr++ = pv[0]; *ptr++ = pv[1]; *ptr++ = pv[2];}
	    }
	}
    }
  if((flag & EZ_DEPTH_BUFFER_BIT) && (EZ_ZBufferSet != 0))
   memset(EZ_Zbuf,0, ((EZ_WindowWidth * EZ_WindowHeight) * sizeof(int)));
}
/***************************************************************************/



